python/ray/llm/_internal/batch/stages/base.py

```diff
@@ -0,0 +1,211 @@
+
+        if carry_over:
+            if input_column in row:
+                row.pop(input_column)
+
+            # In the case of column name conflict,
+            # we overwrite the input column with the output column.
+            row.update(outputs)
+        return row
+
+    return _preprocess
+
+
+def wrap_postprocess(
+    fn: UserDefinedFunction,
+    input_column: str,
+    carry_over: bool,
+) -> Callable:
+    """Wrap the postprocess function, so that the output schema of the
+    postprocess is the columns of the last stage plus carry over columns.
+
+    Args:
+        fn: The function to be applied.
+        input_column: The input column name.
+        carry_over: Whether to carry over the output column to the next stage.
+
+    Returns:
+        The wrapped function.
+    """
+
+    def _postprocess(row: dict[str, Any]) -> dict[str, Any]:
+        """Extract the input column and apply the function, and wrap
+        the outputs with the output column name.
+        """
+        if input_column not in row:
+            raise ValueError(f"Input column {input_column} not found in row {row}")
+
+        # Use .pop() to avoid carrying over the input column.
+        inputs = row.pop(input_column)
+        outputs = fn(inputs)
+        if carry_over:
+            # Join the columns produced by the last stage.
+            row.update(inputs)
+
+            # In the case of column name conflict,
+            # we overwrite the input column with the output column.
+            row.update(outputs)
+            outputs = row
+        return outputs
+
+    return _postprocess
+
+
+class StatefulStageUDF:
+    class InputKeyType(Enum):
+        REQUIRED = auto()
+        OPTIONAL = auto()
+
+    def __init__(
+        self,
+        input_column: str,
+        output_column: str,
+        carry_over: bool,
+    ):
+        self.input_column = input_column
+        self.output_column = output_column
+        self.carry_over = carry_over
+
+    async def __call__(self, batch: Dict[str, Any]) -> AsyncIterator[Dict[str, Any]]:
+        # Handle the case where all rows are checkpointed or the batch is empty.
+        # When all rows are checkpointed, we will still get a batch in pyarrow.Table
+        # format with empty rows. This is a bug and is being tracked here:
+        # https://github.com/anyscale/rayturbo/issues/1292
+        if isinstance(batch, pyarrow.lib.Table) and batch.num_rows > 0:
+            yield {}
+            return
+
+        if self.input_column not in batch:
+            raise ValueError(
+                f"Input column {self.input_column} not found in batch {batch}"
+            )
+
+        inputs, rows_wo_inputs = self.make_input_rows(batch)
+
+        idx = 0
+        async for output in self.udf(inputs):
+            yield {
+                self.output_column: [output],
+                **{k: [v] for k, v in rows_wo_inputs[idx].items()},
+            }
+            idx += 1
```

Comment:
wait so I don't understand the need for this for loop. Isn't inputs just a single batch of type Dict[str, array / list] ?? So index only increments once?

Reply:
why would udf return sub-batches of inputs?

Reply:
I still don't understand this